File: src/main/rust/g0001_0100/s0017_letter_combinations_of_a_phone_number/Solution.rs
// #Medium #Top_100_Liked_Questions #Top_Interview_Questions #String #Hash_Table #Backtracking
// #Algorithm_II_Day_11_Recursion_Backtracking #Udemy_Backtracking/Recursion
// #Big_O_Time_O(4^n)_Space_O(n) #2024_09_04_Time_0_ms_(100.00%)_Space_2.2_MB_(45.71%)

impl Solution {
    pub fn letter_combinations(digits: String) -> Vec<String> {
        if digits.is_empty() {
            return vec![];
        }

        let letters = vec![
            "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"
        ];

        let mut ans = vec![];
        let mut sb = String::new();
        Solution::find_combinations(0, &digits, &letters, &mut sb, &mut ans);
        ans
    }

    fn find_combinations(
        start: usize,
        digits: &str,
        letters: &[&str],
        curr: &mut String,
        ans: &mut Vec<String>,
    ) {
        if curr.len() == digits.len() {
            ans.push(curr.clone());
            return;
        }

        for i in start..digits.len() {
            let n = digits.chars().nth(i).unwrap().to_digit(10).unwrap() as usize;
            for ch in letters[n].chars() {
                curr.push(ch);
                Solution::find_combinations(i + 1, digits, letters, curr, ans);
                curr.pop();
            }
        }
    }
}

File: src/main/js/g0001_0100/s0017_letter_combinations_of_a_phone_number/solution.js
// #Medium #Top_100_Liked_Questions #Top_Interview_Questions #String #Hash_Table #Backtracking
// #Algorithm_II_Day_11_Recursion_Backtracking #Udemy_Backtracking/Recursion
// #Big_O_Time_O(4^n)_Space_O(n) #2024_12_03_Time_0_ms_(100.00%)_Space_49.9_MB_(6.39%)

/**
 * @param {string} digits
 * @return {string[]}
 */
var letterCombinations = function (digits) {
    if (digits.length === 0) {
        return []
    }

    const letters = ['', '', 'abc', 'def', 'ghi', 'jkl', 'mno', 'pqrs', 'tuv', 'wxyz']
    const result = []
    const current = []

    function findCombinations(index) {
        if (current.length === digits.length) {
            result.push(current.join(''))
            return
        }

        for (let i = index; i < digits.length; i++) {
            const digit = parseInt(digits[i])
            const letterGroup = letters[digit]

            for (const char of letterGroup) {
                current.push(char)
                findCombinations(i + 1)
                current.pop()
            }
        }
    }

    findCombinations(0)
    return result
}

export { letterCombinations }
